src/service/resource.js

```diff
@@ -654,51 +661,61 @@ export class Resource {
   }
 
   /**
+   * @param {number|boolean} viewport derived from renderOutsideViewport.
    * @return {!Promise} resolves when underlying element is built and within the
-   *    range specified by implementation's renderOutsideViewport
+   *    viewport range given.
    */
-  whenWithinRenderOutsideViewport() {
-    if (!this.isLayoutPending()) {
+  whenWithinViewport(viewport) {
+    dev().assert(viewport !== false);
+    // Resolve is already laid out or viewport is true.
+    if (!this.isLayoutPending() || viewport === true) {
       return Promise.resolve();
     }
-    if (this.renderOutsideViewportPromise_) {
-      return this.renderOutsideViewportPromise_;
+    // See if pre-existing promise.
+    const viewportNum = dev().assertNumber(viewport);
+    if (this.withViewportDeferreds_ &&
+        this.withViewportDeferreds_[viewportNum]) {
+      return this.withViewportDeferreds_[viewportNum].promise;
     }
-    const deferred = new Deferred();
-    this.renderOutsideViewportResolve_ = deferred.resolve;
-    return this.renderOutsideViewportPromise_ = deferred.promise;
+    // See if already within viewport multiplier.
+    if (this.isWithinViewportRatio(viewport)) {
+      return Promise.resolve();
+    }
+    // return promise that will trigger when within viewport multiple.
+    this.withViewportDeferreds_ = this.withViewportDeferreds_ || {};
+    this.withViewportDeferreds_[viewportNum] = new Deferred();
+    return this.withViewportDeferreds_[viewportNum].promise;
   }
 
-  /**
-   * @private resolves render outside viewport promise if one was created via
-   *    whenWithinRenderOutsideViewport.
-   */
-  resolveRenderOutsideViewport_() {
-    if (!this.renderOutsideViewportResolve_) {
+  /** @private resolves promises populated via whenWithinViewport. */
+  resolveDeferredsWhenWithinViewports_() {
+    if (!this.withViewportDeferreds_) {
       return;
     }
-    this.renderOutsideViewportResolve_();
-    this.renderOutsideViewportPromise_ = null;
-    this.renderOutsideViewportResolve_ = null;
-  }
-
-  /**
-   * @param {number|boolean} multiplier
-   * @return {boolean} whether resource is within provider multiplier of
-   *    viewports from current visible viewport.
-   * @visibleForTesting
-   */
-  withinViewportMultiplier(multiplier) {
-    // Boolean interface controls explicit result.
-    if (multiplier === true || multiplier === false) {
-      return multiplier;
+    const viewportRatio = this.getDistanceViewportRatio();
+    const keys = Object.keys(this.withViewportDeferreds_);
+    let removed = 0;
+    keys.forEach(viewport => {
+      const viewportInt = dev().assertNumber(parseInt(viewport, 10));
+      if (this.isWithinViewportRatio(viewportInt, viewportRatio)) {
+        this.withViewportDeferreds_[viewportInt].resolve();
+        delete this.withViewportDeferreds_[viewportInt];
+        removed++;
+      }
+    });
+    if (removed == keys.length) {
+      this.withViewportDeferreds_ = null;
```

Comment:
Nit: it's unlikely this will really help with memory savings. I'm ok with leaving the map in place if it's ever been needed.